node/src/Pos/Update/Download.hs

```diff
@@ -22,80 +21,74 @@ import           Network.HTTP.Simple     (getResponseBody, getResponseStatus,
-import           System.Wlog             (logDebug, logInfo, logWarning)
-import           Universum
+import           System.Wlog             (logInfo, logWarning)
 
 import           Pos.Binary.Update       ()
 import           Pos.Constants           (appSystemTag, curSoftwareVersion)
 import           Pos.Core.Types          (SoftwareVersion (..))
 import           Pos.Crypto              (Hash, castHash, hash)
+import           Pos.Exception           (reportFatalError)
 import           Pos.Update.Context      (UpdateContext (..))
 import           Pos.Update.Core.Types   (UpdateData (..), UpdateProposal (..))
 import           Pos.Update.Mode         (UpdateMode)
 import           Pos.Update.Params       (UpdateParams (..))
 import           Pos.Update.Poll.Types   (ConfirmedProposalState (..))
 import           Pos.Util                ((<//>))
-
-showHash :: Hash a -> FilePath
-showHash = toString . B16.encode . BA.convert
-
--- CSL-887: if we're downloading update not for `cardano-sl`,
--- but e. g. for `daedalus`, how do we check that version is new?
-versionIsNew :: SoftwareVersion -> Bool
-versionIsNew ver = svAppName ver /= svAppName curSoftwareVersion
-    || svNumber ver > svNumber curSoftwareVersion
-
--- TODO Now we suppose there is no more than one update at every moment.
--- | Determine whether to download update and download it if needed.
+import           Pos.Util.Concurrent     (withMVar)
+
+-- | Download a software update for given 'ConfirmedProposalState' and
+-- put it into a variable which holds 'ConfirmedProposalState' of
+-- downloaded update.
+-- Parallel downloads aren't supported, so this function may blocks.
+-- If we have already downloaded an update successfully, this function won't
+-- download new updates.
 downloadUpdate :: forall ctx m . UpdateMode ctx m => ConfirmedProposalState -> m ()
-downloadUpdate cst@ConfirmedProposalState {..} = do
-    unlessM (liftIO . doesFileExist =<< views (lensOf @UpdateParams) upUpdatePath) $ do
-        downSetVar <- views (lensOf @UpdateContext) ucDownloadingUpdates
-        (forDownload, downSet) <- tryPutToSet downSetVar cpsUpdateProposal
-        if forDownload
-           then do
-              logDebug $ sformat ("Update downloading triggered, download state: "%listJson)
-                                 downSet
-              downloadUpdateDo cst
-                `finally` clear downSetVar cpsUpdateProposal
-           else
-              logDebug $ sformat ("Update downloading already in progress, download state: "%listJson)
-                                 downSet
-
+downloadUpdate cps = do
+    downloadLock <- ucDownloadLock <$> view (lensOf @UpdateContext)
+    withMVar downloadLock $ \() -> do
+        downloadedUpdateMVar <-
+            ucDownloadedUpdate <$> view (lensOf @UpdateContext)
+        tryTakeMVar downloadedUpdateMVar >>= \case
+            Nothing -> downloadUpdateDo cps
+            Just existingCPS -> onAlreadyDownloaded existingCPS
```

Comment:
Why we take `MVar` here, log something and don't put it back? Perhaps Tvar should be used

Reply:
Fix already.

Reply:
Right